test/testcafe/spec/ChallengeAuthenticatorDuo_spec.js

```diff
@@ -0,0 +1,86 @@
+import { RequestMock, RequestLogger } from 'testcafe';
+
+import SuccessPageObject from '../framework/page-objects/SuccessPageObject';
+import DuoPageObject from '../framework/page-objects/DuoPageObject';
+import xhrAuthenticatorVerifyDuo from '../../../playground/mocks/data/idp/idx/authenticator-verification-duo';
+import success from '../../../playground/mocks/data/idp/idx/success';
+
+const mock = RequestMock()
+  .onRequestTo('http://localhost:3000/idp/idx/introspect')
+  .respond(xhrAuthenticatorVerifyDuo)
+  .onRequestTo('http://localhost:3000/idp/idx/challenge/answer')
+  .respond(success);
+
+const answerRequestLogger = RequestLogger(
+  /idx\/challenge\/answer/,
+  {
+    logRequestBody: true,
+    stringifyRequestBody: true,
+  }
+);
+
+fixture('Challenge Duo');
+
+async function setup(t) {
+  const challengeDuoPage = new DuoPageObject(t);
+  await challengeDuoPage.navigateToPage();
+
+  const { log } = await t.getBrowserConsoleMessages();
+  await t.expect(log.length).eql(3);
+  await t.expect(log[0]).eql('===== playground widget ready event received =====');
+  await t.expect(log[1]).eql('===== playground widget afterRender event received =====');
```

Comment:
@haishengwu-okta - This code is duplicated in nearly all testcafe tests. Plus, we shouldn't need to always have expectations for these event logs are triggered.

Couple of suggestions:
- Ask developers to only assert `log[2]` contains the controller information in this `setup` call. The remaining assertions can be either be checked in a shared function or ignored completely
- Look into using [shared test fixtures](https://devexpress.github.io/testcafe/documentation/guides/basic-guides/organize-tests.html#share-variables-between-test-hooks-and-test-code) for validating common code.

Reply:
either makes sense. Go for any improvement to avoid redundancy!

Reply:
Created https://oktainc.atlassian.net/browse/OKTA-369480 to follow up